design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,347 @@
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDFs.
+
+## Design
+
+### General plumbing
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDFs are currently supported we will need to create a 
+`KsqlTableFunction` which is a wrapper around the actual table function (whether built-in or
+generated). We'll also need to augment the function registry to manage the addition and retrieval
+of table functions.
+* Augment the query analysis code to extract information relating to any table functions
+in the query. We can do this either by extending the existing query analyzer or perhaps creating
+a new `TableFunctionAnalyzer` in a similar way to how aggregate analysis uses its own 
+`AggregateAnalyzer`.
+* Augment the building of logical plan by building a new node representing a FlatMap stage of the
+stream processing in the case there are table functions in the SELECT.
+* Augment the execution plan logic by adding new nodes in the execution plan corresponding to any
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
+The number of rows returned from the FlatMap will be equal to the greatest of the number of rows
+returned from any of the table functions for that row. Any non table function values for the row will
+be duplicated for each row. For any table function that returned fewer than the greatest number of rows
+the values for that columnn will be `null`. For performance reasons, the FlatMap logic should cache any
+ table functions for the columns to avoid them being looked up each time.
+ 
+### UDTF API
+
+Users will be able to implement their own table function by creating their classes containing
+methods annotated with `@Udtf`.
+
+A UDTF method can contain zero or more parameters, of any of the valid KSQL types mapped to Java.
+The return value must be a `List<T>` where T is any valid KSQL type mapped to Java. 
+
+```
+@UdtfDescription(name = "SplitSpring", author = "bob", description = "Splits a string into words")
+public class SplitString {
+
+    @Udtf(description="Splits a string into words")
+    public List<String> split(String input) {
+        return Arrays.asList(String.split("\\s+"));
+    }
+
+}
+
+```
+
+### Implement set of built in table functions
+
+We will implement the following built in table functions:
+
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `SORTED_PAIRS(MAP<K, V>)` -> Results in `N` rows (where `N` is number of entries in map), sorted by `K`
```

Comment:
This is an interesting point. If we take this route then we only ever need an EXPLODE table function, all the other things that could otherwise be table functions could be implemented as normal UDFs which return a List<T>. E.g. splitting a string into words, or series().

We could go either route, but if we go the route of only using table functions to explode (that's the only thing a udf can't do), then we could disallow users from creating their own table functions (because they don't need, they can write a udf instead).

Reply:
I guess, but allowing custom UDTFs allows for users to make it easier / more succinct to write their SQL.  This is a good thing.

Reply:
On the basis that SORTED_PAIRS can just be a UDF, the same logic also applies to SERIES and UNSORTED_PAIRS, so I'm going to remove these functions from the KLIP, and we'll only implement EXPLODE (we'll keep the UDTF functionality in though).